File: source/solution_deploy/source/cfnresponse.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
"""Send custom resource status to CloudFormation"""

import json

import urllib3

SUCCESS = "SUCCESS"
FAILED = "FAILED"

http = urllib3.PoolManager()


def send(
    event,
    context,
    response_status,
    response_data,
    physical_resource_id=None,
    no_echo=False,
    reason=None,
):
    """Send custom resource status to CloudFormation"""
    response_url = event["ResponseURL"]

    print(response_url)

    max_reason_length = 3854  # response can't exceed 4 kiB
    if reason and len(reason) > max_reason_length:
        reason = reason[:max_reason_length]

    response_body = {
        "Status": response_status,
        "Reason": reason
        or f"See the details in CloudWatch Log Stream: {context.log_stream_name}",
        "PhysicalResourceId": physical_resource_id or context.log_stream_name,
        "StackId": event["StackId"],
        "RequestId": event["RequestId"],
        "LogicalResourceId": event["LogicalResourceId"],
        "NoEcho": no_echo,
        "Data": response_data,
    }

    json_response_body = json.dumps(response_body)

    print("Response body:")
    print(json_response_body)

    headers = {"content-type": "", "content-length": str(len(json_response_body))}

    try:
        response = http.request(
            "PUT", response_url, headers=headers, body=json_response_body
        )
        print("Status code:", response.status)

    except Exception as ex:
        print("send(..) failed executing http.request(..):", ex)

File: source/lambdas/api/handlers/cfnResponse.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as https from 'https';
import { CloudFormationCustomResourceEvent, Context } from 'aws-lambda';

export const SUCCESS = 'SUCCESS';
export const FAILED = 'FAILED';

// copied from https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html
// and converted from JS to TS
export function send(
  event: CloudFormationCustomResourceEvent,
  context: Context,
  responseStatus: typeof SUCCESS | typeof FAILED,
  responseData: Record<string, any>,
  physicalResourceId?: string,
  noEcho?: boolean,
): Promise<void> {
  return new Promise((resolve, reject) => {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason: 'See the details in CloudWatch Log Stream: ' + context.logStreamName,
      PhysicalResourceId: physicalResourceId || context.logStreamName,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: noEcho || false,
      Data: responseData,
    });

    console.log('Response body:\n', responseBody);

    const parsedUrl = new URL(event.ResponseURL);
    const options: https.RequestOptions = {
      hostname: parsedUrl.hostname,
      port: 443,
      path: parsedUrl.pathname + parsedUrl.search,
      method: 'PUT',
      headers: {
        'content-type': '',
        'content-length': responseBody.length,
      },
    };

    const request = https.request(options, (response) => {
      console.log('Status code: ' + response.statusCode);
      resolve();
    });

    request.on('error', (error) => {
      console.log('send(..) failed executing https.request(..): ' + maskCredentialsAndSignature(error.message));
      reject(error);
    });

    request.write(responseBody);
    request.end();
  });
}

function maskCredentialsAndSignature(message: string): string {
  return message
    .replace(/X-Amz-Credential=[^&\s]+/i, 'X-Amz-Credential=*****')
    .replace(/X-Amz-Signature=[^&\s]+/i, 'X-Amz-Signature=*****');
}
